### PR TITLE
Now with bunnies.

### PR DESCRIPTION
**Pull request by peterkagey**
If these bunnies don't work for you, I don't even...

### CODE REVIEW

File: lib/cell.rb [hunk 1]
```diff
@@ -1,33 +1,31 @@
 class Cell
   attr_accessor :state
 
-  def initialize(state = nil)
-    @state = state || set_random_state
+  DEAD_CELL = Cell.new
+  ALIVE_CELL = Cell.new
```
Comment (peterkagey) [inline lib/cell.rb:5]: I really like how this shows the intent.

File: lib/grid.rb [hunk 2]
```diff
@@ -1,5 +1,6 @@
 class Grid
-  DEFAULT_WIDTH = 50
+
+  DEFAULT_WIDTH = `tput cols`.to_i/2 - 4
```
Comment (peterkagey) [inline lib/grid.rb:3]: This autoscales to the size of the terminal.

File: lib/grid.rb [hunk 3]
```diff
@@ -15,25 +16,21 @@ def initialize(width = DEFAULT_WIDTH, height = DEFAULT_HEIGHT, cells = nil)
...
 
   def next_cells
-    _next_cells = {}
-    cells.each do |y, rows|
-      new_row = {}
-      rows.each do |x, cell|
-        new_row[x] = next_cell_for(cell, x, y)
+    cells.each_with_index.map do |rows, y|
+      rows.each_with_index.map do |cell, x|
+        next_cell_for(cell, x, y)
```
Comment (peterkagey) [inline lib/grid.rb:31]: Store this as an array of arrays—not a hash.

File: lib/grid.rb [hunk 4]
```diff
@@ -59,27 +56,21 @@ def survives?(cell, x, y)
   def alive_neighbors(x, y)
     neighbors = []
 
-    neighbors << cells[x-1][y]   rescue nil
+    neighbors << cells[x-1][y]   unless x == 0
```
Comment (peterkagey) [inline lib/grid.rb:59]: Let's not have the `x-1` and `y-1` indices wrap.

File: lib/presenter.rb [hunk 5]
```diff
@@ -5,20 +5,13 @@ def initialize(grid)
...
 
-  def print
-    STDOUT.print "\n"
-    grid.height.times do |y|
-      grid.width.times do |x|
-        print_cell(grid.cells[y][x])
-      end
-      STDOUT.print "\n"
-    end
+  def print_grid
```
Comment (peterkagey) [inline lib/presenter.rb:8]: Instead of printing out each character individually, only print fully-formed frames.

File: spec/presenter_spec.rb [hunk 6]
```diff
@@ -1,19 +1,23 @@
...
+      [d, d, a],
+    ]
 
     grid = Grid.new(3, 3, cells)
-    expect { grid.presenter.print }.to output("•• \n•• \n•• \n").to_stdout
+    s = Cell.alive.to_s
+    expected_print = [
+      "🐰 🐰  ",
+      "🐰   🐰",
+      "    🐰",
```
Comment (peterkagey) [inline spec/presenter_spec.rb:17]: These bunnies are too wide for their own good.

**Comment by peterkagey** [comment]
It's a masterpiece!

File: lib/grid.rb (hunk 2 above)
Comment (mzemel) [reply lib/grid.rb:3]: This is fancy but is it portable to Windows? We are expecting a large MS demographic to buy our product soon.

File: spec/presenter_spec.rb (hunk 6 above)
Comment (mzemel) [reply spec/presenter_spec.rb:17]: Will this be able to be stored in a UTF8 database column? We've committed to supported MySQL 5.4 for another 6 months.

File: lib/grid.rb [hunk 7]
```diff
@@ -15,25 +16,21 @@ def initialize(width = DEFAULT_WIDTH, height = DEFAULT_HEIGHT, cells = nil)
...
-    presenter.print
+    presenter.print_grid
   end
 
   def next_grid
     Grid.new(width, height, next_cells)
   end
 
-  private
+  # private
```
Comment (mzemel) [inline lib/grid.rb:26]: Is this method being used outside of the class?

File: lib/grid.rb [hunk 8]
```diff
@@ -59,27 +56,19 @@ def survives?(cell, x, y)
...
-    neighbors << cells[x-1][y-1] rescue nil
-    neighbors << cells[x+1][y-1] rescue nil
-    neighbors << cells[x-1][y+1] rescue nil
+    neighbors << cells[x-1][y-1] unless x == 0 || y == 0
+    neighbors << cells[x+1][y-1] unless y == 0 rescue nil
+    neighbors << cells[x-1][y+1] unless x == 0 rescue nil
     neighbors << cells[x+1][y+1] rescue nil
-    neighbors << cells[x][y-1]   rescue nil
+    neighbors << cells[x][y-1]   unless y == 0
     neighbors << cells[x][y+1]   rescue nil
```
Comment (mzemel) [inline lib/grid.rb:67]: I hate everything about the code I wrote here

File: lib/presenter.rb [hunk 9]
```diff
@@ -5,20 +5,13 @@ def initialize(grid)
...
+    print grid.cells.map { |row| row_string(row) }.join("\n")
   end
 
   private
 
-  def print_cell(cell)
-    val = cell.alive? ? '• ' : '  '
```
Comment (mzemel) [inline lib/presenter.rb:21]: I think the point of this presenter was to take the view logic out of the cell model. But fuck it dawg, it's Friday night.

**Comment by mzemel** [comment]
Merged into master locally.

File: spec/presenter_spec.rb (hunk 6 above)
Comment (peterkagey) [reply spec/presenter_spec.rb:17]: Yep. Every unicode character has a UTF-8 encoding.

**Pull request closed without merging**